[clang] Add builtin_get_vtable_pointer and virtual_member_address

ojhunt opened the pull request:
These are a pair of builtins to support particularly weird edge case operations while correctly handling the non-trivial implicit pointer authentication schemas applied to polymorphic members.

Co-authored-by: Tim Northover

asl commented on clang/test/CodeGenCXX/builtin_virtual_member_address.cpp:
```diff
@@ -0,0 +1,41 @@
+// RUN: %clang_cc1 -triple arm64e-apple-ios -fptrauth-calls -emit-llvm -no-enable-noundef-analysis -o - %s | FileCheck %s --check-prefixes CHECK,CHECK-AUTH
+// RUN: %clang_cc1 -triple arm64-apple-ios -emit-llvm -no-enable-noundef-analysis -o - %s | FileCheck %s --check-prefixes CHECK,CHECK-NOAUTH
```
will you please also add ELF target triples in tests?

ojhunt replied:
Sure thing, what is the triple you folk use?

ojhunt replied:
@asl :D

ojhunt replied:
From discord: aarch64-linux-pauthtest

ojhunt commented:
I just noticed there's no documentation for `__builtin_virtual_member_address()` so I'm addressing that.

Due to code drift and time I realized that they disagree as to whether they should take `C*` or `C&` so I'm going to make them both accept either (existing code means we can't unify on one or the other, but also there's no real reason to not accept either and simply do the correct thing)

ojhunt committed:
[clang] Add builtin_get_vtable_pointer and virtual_member_address

These are a pair of builtins to support particularly weird edge
case operations while correctly handling the non-trivial implicit
pointer authentication schemas applied to polymorphic members.

Co-authored-by: Tim Northover

cor3ntin commented:
> I just noticed there's no documentation for `__builtin_virtual_member_address()` so I'm addressing that.
> 
> Due to code drift and time I realized that they disagree as to whether they should take `C*` or `C&` so I'm going to make them both accept either (existing code means we can't unify on one or the other, but also there's no real reason to not accept either and simply do the correct thing)

most of our builtins take pointers. Accepting both seems unwise.
people might want to do interesting things in sfinae context and if we accept everything and anything it will end up causing more pain than it alleviates.

ojhunt commented:
> > I just noticed there's no documentation for `__builtin_virtual_member_address()` so I'm addressing that.
> > Due to code drift and time I realized that they disagree as to whether they should take `C*` or `C&` so I'm going to make them both accept either (existing code means we can't unify on one or the other, but also there's no real reason to not accept either and simply do the correct thing)
> 
> most of our builtins take pointers. Accepting both seems unwise. people might want to do interesting things in sfinae context and if we accept everything and anything it will end up causing more pain than it alleviates.

Hmmm, I'll revert to only supporting pointers in ``get_vtable_pointer`` but there's existing code dependent on the non-pointer argument to virtual_member_address which makes it difficult to switch to having a pointer argument.

ojhunt commented:
> > I just noticed there's no documentation for `__builtin_virtual_member_address()` so I'm addressing that.
> > Due to code drift and time I realized that they disagree as to whether they should take `C*` or `C&` so I'm going to make them both accept either (existing code means we can't unify on one or the other, but also there's no real reason to not accept either and simply do the correct thing)
> 
> most of our builtins take pointers. Accepting both seems unwise. people might want to do interesting things in sfinae context and if we accept everything and anything it will end up causing more pain than it alleviates.

@AaronBallman Given we need to keep the reference argument to __virtual_member_address, would you like to support both, or remain as reference only?

AaronBallman commented:
> > > I just noticed there's no documentation for `__builtin_virtual_member_address()` so I'm addressing that.
> > > Due to code drift and time I realized that they disagree as to whether they should take `C*` or `C&` so I'm going to make them both accept either (existing code means we can't unify on one or the other, but also there's no real reason to not accept either and simply do the correct thing)
> > 
> > 
> > most of our builtins take pointers. Accepting both seems unwise. people might want to do interesting things in sfinae context and if we accept everything and anything it will end up causing more pain than it alleviates.
> 
> Hmmm, I'll revert to only supporting pointers in `get_vtable_pointer` but there's existing code dependent on the non-pointer argument to virtual_member_address which makes it difficult to switch to having a pointer argument.

I'm confused -- how is there existing code depending on a builtin which doesn't yet exist in Clang? Are we copying these builtins from somewhere else?

ojhunt commented:
> I'm confused -- how is there existing code depending on a builtin which doesn't yet exist in Clang? Are we copying these builtins from somewhere else?

This is part of the pointer authentication support we are upstreaming, all of which has been in use for a long time, just not available in upstream llvm/clang.

AaronBallman commented:
> > I'm confused -- how is there existing code depending on a builtin which doesn't yet exist in Clang? Are we copying these builtins from somewhere else?
> 
> This is part of the pointer authentication support we are upstreaming, all of which has been in use for a long time, just not available in upstream llvm/clang.

Oof. On the one hand, we don't want to break your existing users. On the other hand, we don't want to force the much broader audience of users into a worse interface. Are those uses in your downstream able to use things like fix-its to switch from the reference-based interface to a pointer-based interface? (Basically, is there a transition approach that leads to a better world?)

shafik commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -5349,6 +5350,40 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+  case Builtin::BI__builtin_virtual_member_address: {
+    Address This = EmitLValue(E->getArg(0)).getAddress();
+    APValue ConstMemFun;
+    E->getArg(1)->isCXX11ConstantExpr(getContext(), &ConstMemFun, nullptr);
```
So we don't care about the return value of this call? I guess b/c we verify in `VirtualMemberAddress`, so we are using this for a side effect?

ojhunt replied:
Yes, the isCXX11ConstantExpr gives us the target through ConstMemFun, but I'll see if there's a cleaner way to do this.

cor3ntin replied:
`Expr::EvaluateAsLValue` ?

shafik commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -5349,6 +5350,40 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+    if (RecordType->isPointerOrReferenceType())
+      RecordType = RecordType->getPointeeType();
+    const CXXRecordDecl *Decl = RecordType->getAsCXXRecordDecl();
+    assert(Decl);
```
I note you are asserting pointers here but not in `BI__builtin_virtual_member_address` case, not obvious to me why.

ojhunt replied:
Different authors over time, I'll unify on not asserting as null deref is the downstream outcome.

ojhunt commented:
> Oof. On the one hand, we don't want to break your existing users. On the other hand, we don't want to force the much broader audience of users into a worse interface. Are those uses in your downstream able to use things like fix-its to switch from the reference-based interface to a pointer-based interface? (Basically, is there a transition approach that leads to a better world?)

I sent this in discord rather than in GH where the reply can be seen by others and not lost to time:

> i'm going to revert the addition of reference parameter i added to __builtin_get_vtable_pointer, for the suboptimal behavior of __builtin_virtual_member_address, i could make it pointer argument only by default, with a flag to permit reference parameters and a deprecation warning so that it's only no new users of a reference argument arrive and we can drop it entirely once all existing users have migrated

ojhunt commented:
pending the above changes I'll remove the review flag so people aren't confused

ojhunt commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -1782,6 +1782,92 @@ static ExprResult PointerAuthStringDiscriminator(Sema &S, CallExpr *Call) {
+  Call->setArg(0, ThisArg.get());
+  const Expr *Subject = Call->getArg(0);
+  QualType SubjectType = Subject->getType();
+  if (SubjectType->isPointerOrReferenceType())
```
Will be reverting this to require a pointer arg

ojhunt commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -1782,6 +1782,92 @@ static ExprResult PointerAuthStringDiscriminator(Sema &S, CallExpr *Call) {
+
+  const Expr *ThisArg = Call->getArg(0);
+  QualType ThisTy = ThisArg->getType();
+  if (ThisTy->isPointerOrReferenceType())
```
I'll update this to make it warn on a reference parameter

ojhunt commented:
Closing this, as I'm not sure the virtual_member_address builtin is actually needed, so for now we'll just do `builtin_get_vtable` which is used in other projects
